### PR TITLE
Feature/calculateauc

### PR DESCRIPTION
**Pull request by antschum**
**Description of changes**
Implementation to calculate the AUC with respect to CellType along with test files. 

Still missing right now: 
-  is a nice results output. (Current version returns huge dictionary with everything..)
- feature importances (still not sure 1. how to best return them - eg. do I take the average if multiple importances are captured and 2. feature importances for logistic regression -> Issue.)
- progress bar for calculations

### CODE REVIEW

File: augurpy/calculate_AUC.py [hunk 1]
```diff
@@ -1,66 +1,70 @@
...
-        subsample_size: number of cells to subsample randomly per type from each experimental condition
-        folds: number of folds to run cross validation on
-        min_cells: minimum number of cells for a particular cell type in each condition in order to retain that type for
-            analysis (depricated..)
-        var_quantile: quantile of highly variable genes to retain for each cell type using the variable gene filter
-        feature_perc: proportion of genes that are randomly selected as features for input to the classifier in each
-            subsample using the random gene filter
-        n_threads: number of threads to use for parallelization
-        show_progress: if `True` display a progress bar for the analysis with estimated time remaining
+        input: Pandas DataFrame
```
Comment (Zethson) [inline augurpy/calculate_AUC.py]: Doesn't quite match the input type, right?

File: augurpy/calculate_AUC.py [hunk 2]
```diff
@@ -1,66 +1,70 @@
...
+        estimator: classifier
```
Comment (Zethson) [inline augurpy/calculate_AUC.py]: classifier doing what? :)

File: augurpy/calculate_AUC.py [hunk 3]
```diff
@@ -1,66 +1,70 @@
...
             assuming feature selection has been performed by the RNA velocity procedure to produce the input matrix,
             while setting augur_mode = "permute" will generate a null distribution of AUCs for each cell type by
             permuting the labels
-        classifier: classifier to use in calculating the area under the curve either random forest or logistic regression
-        rf_params: list of parameters for random forest
-        lr_params: list of parameters for logistic regression
+        subsample_size: number of cells to subsample randomly per type from each experimental condition
+        folds: number of folds to run cross validation on
+        subsample_idx: index of the subsample
+        random_state: set random seed
```
Comment (Zethson) [inline augurpy/calculate_AUC.py]: What does the random seed effect? Very briefly mention here. Note if it does not affect something it might (e.g. Numpy seed or whatever)

File: augurpy/calculate_AUC.py [hunk 4]
```diff
@@ -1,66 +1,70 @@
...
+    augur_mode: str,
+    subsample_size: int,
+    feature_perc: float,
+    stratified: bool,
+    random_state: Optional[int],
+) -> AnnData:
+    """Subsample input and select random features.
 
     Args:
         input: Pandas DataFrame containing gene expression values (cells in rows, genes in columns) along with cell type and
```
Comment (Zethson) [inline augurpy/calculate_AUC.py]: Also doesn't match the input type.

File: augurpy/calculate_AUC.py [hunk 5]
```diff
@@ -71,39 +75,110 @@ def draw_subsample(input: DataFrame, augur_mode: str, subsample_size: int, strat
             permuting the labels
         subsample_size: number of cells to subsample randomly per type from each experimental condition
         stratified: if `True` subsamples are stratified according to condition
+        random_state: set random seed
```
Comment (Zethson) [inline augurpy/calculate_AUC.py]: See above.

File: augurpy/calculate_AUC.py [hunk 6]
```diff
@@ -71,39 +75,110 @@ def draw_subsample(input: DataFrame, augur_mode: str, subsample_size: int, strat
...
 
     Returns:
         Subsample of input of size subsample_size
     """
-    pass
+    if augur_mode == "permut":
+        # shuffle labels
+        y_columns = [col for col in adata.obs if col.startswith("y")]
+        adata.obs[y_columns] = adata.obs[y_columns].sample(frac=1, random_state=random_state).reset_index(drop=True)
```
Comment (Zethson) [inline augurpy/calculate_AUC.py:87]: Why is it required to reset the index here?

File: augurpy/calculate_AUC.py [hunk 7]
```diff
@@ -71,39 +75,110 @@ def draw_subsample(input: DataFrame, augur_mode: str, subsample_size: int, strat
...
-    stratified: bool,
-    estimator: Union[RandomForestRegressor, RandomForestClassifier, LogisticRegression],
-    metrics: list,
-    folds: int,
-    subsample_idx: int,
-) -> DataFrame:
-    """Cross validate subsample.
+    # randomly sample features
+    # random.seed()
+    print(random, type(random))
```
Comment (Zethson) [inline augurpy/calculate_AUC.py]:
```suggestion
```

File: augurpy/calculate_AUC.py [hunk 8]
```diff
@@ -71,39 +75,110 @@ def draw_subsample(input: DataFrame, augur_mode: str, subsample_size: int, strat
...
-    subsample_size: int,
-    stratified: bool,
-    estimator: Union[RandomForestRegressor, RandomForestClassifier, LogisticRegression],
-    metrics: list,
-    folds: int,
-    subsample_idx: int,
-) -> DataFrame:
-    """Cross validate subsample.
+    # randomly sample features
+    # random.seed()
```
Comment (Zethson) [inline augurpy/calculate_AUC.py]: Should be affected by the passed seed, no? Also, maybe wrap the last block into an `else` block?

File: augurpy/calculate_AUC.py [hunk 9]
```diff
@@ -71,39 +75,110 @@ def draw_subsample(input: DataFrame, augur_mode: str, subsample_size: int, strat
...
+    """Calculate average scores across multiple cross validation runs.
 
     Args:
-        input: Pandas DataFrame
+        cross_validation_results: list of all subsample cross validations
+
+    Returns:
+        Dict containing the average result for each metric.
+    """
+    metric_names = [m for m in [*subsample_cv_results[0].keys()] if m.startswith("mean")]
```
Comment (Zethson) [inline augurpy/calculate_AUC.py]:
```suggestion
    metric_names = [metric for metric in [*subsample_cv_results[0].keys()] if metric.startswith("mean")]
```

As long as it stays readable always prefer explicit over short variable names.

File: augurpy/calculate_AUC.py [hunk 10]
```diff
@@ -71,39 +75,110 @@ def draw_subsample(input: DataFrame, augur_mode: str, subsample_size: int, strat
...
+    metric_list: Dict[Any, Any] = {}
+    for d in subsample_cv_results:
```
Comment (Zethson) [inline augurpy/calculate_AUC.py]: What is `d`? Be explicit, please.

File: augurpy/calculate_AUC.py [hunk 11]
```diff
@@ -71,39 +75,110 @@ def draw_subsample(input: DataFrame, augur_mode: str, subsample_size: int, strat
...
+        for m in metric_names:
```
Comment (Zethson) [inline augurpy/calculate_AUC.py]: m = metric

File: augurpy/cv.py [hunk 12]
```diff
@@ -30,31 +30,36 @@ def run_cross_validation(
...
 ) -> Dict:
     """Perform cross validation on given subsample.
 
     Args:
         subsample: subsample of gene expression matrix of size subsample_size
         estimator: classifier object to use in calculating the area under the curve
-        folds: number of folds
         subsample_idx: index of subsample
+        folds: number of folds
+        random_state: set random seed
```
Comment (Zethson) [inline augurpy/cv.py]: See above.

File: augurpy/cv.py [hunk 13]
```diff
@@ -30,31 +30,36 @@ def run_cross_validation(
...
 
     Returns:
         Dictionary containing prediction metrics and estimator for each fold.
     """
     scorer = set_scorer(estimator)
     x = subsample.to_df()
     y = subsample.obs[[col for col in subsample.obs if col.startswith("y")]]
+    cv = KFold(n_splits=folds, random_state=random_state, shuffle=True)
```
Comment (Zethson) [inline augurpy/cv.py]: Mhm, I liked the variable name folds that you passed before better.

A KFold object is not a cross validation :)

File: augurpy/cv.py [hunk 14]
```diff
@@ -30,31 +30,36 @@ def run_cross_validation(
...
         X=x,
         y=y.values.ravel(),
         scoring=scorer,
-        cv=folds,
+        cv=cv,
         return_estimator=True,
     )
 
     results["subsample_idx"] = subsample_idx
+    for s in scorer.keys():
```
Comment (Zethson) [inline augurpy/cv.py]:
```suggestion
    for score in scorer.keys():
```

File: augurpy/cv.py [hunk 15]
```diff
@@ -30,31 +30,36 @@ def run_cross_validation(
...
+        results[f"mean_{s}"] = results[f"test_{s}"].mean()
```
Comment (Zethson) [inline augurpy/cv.py]:
```suggestion
        results[f"mean_{score}"] = results[f"test_{score}"].mean()
```

**Review by Zethson — COMMENTED** [review]
Cool to see it coming together!

I'd suggest that you rename the Python files to ensure that they are descriptive.
`calculate_AUC.py` for example is not the nicest of all names.

Besides that just many tiny requests by me. Happy to see this merge soon! Then we'll also need to ensure that we properly document all of this. See #40

**Comment by Zethson** [comment]
> feature importances (still not sure 1. how to best return them - eg. do I take the average if multiple importances are captured and 2. feature importances for logistic regression -> Issue.)

Mhmm. Didn't we want to look at how Augur calculates them? Did you figure something out?

**Comment by antschum** [comment]
> > feature importances (still not sure 1. how to best return them - eg. do I take the average if multiple importances are captured and 2. feature importances for logistic regression -> Issue.)
> 
> Mhmm. Didn't we want to look at how Augur calculates them? Did you figure something out?

Augur just a huge matrix containing the raw data like so:
cell type subsample idx fold
dimensions: celltypes x folds x n_subsamples x n genes sampled
Ill implement this for starters, but am not super happy with it.. It would be cool to brainstorm on why this might be interesting to the user to beginn with.

**Comment by Zethson** [comment]
I guess that this should help determine which genes contributed the most to the cell type prioritization?

File: augurpy/cv.py (hunk 13 above)
Comment (antschum) [reply augurpy/cv.py]: done:)

**Comment by Zethson** [comment]
@antschum Laptop working again? ^_^
Should I have another look?

**Comment by antschum** [comment]
> @antschum Laptop working again? ^_^

Hey:) using my brothers, should have a working one by Thursday though. 

> Should I have another look?

That would be great! Ill click on request. 
Still missing feature importances for logistic regression (started reading, I think it sounds more complicated than it is, but don't fully understand it yet. There is not a lot online) and the concordance correlation coefficient for regression models, which I just have to implement.

**Comment by Zethson** [comment]
@antschum think you missed two comments of mine. (You can scroll through the changed files.)  But besides this it looks great! Feel free to merge it after you've addressed my comments.

Regarding the feature importances:

```
# logistic regression for feature importance
from sklearn.datasets import make_classification
from sklearn.linear_model import LogisticRegression
from matplotlib import pyplot
# define dataset
X, y = make_classification(n_samples=1000, n_features=10, n_informative=5, n_redundant=5, random_state=1)
# define the model
model = LogisticRegression()
# fit the model
model.fit(X, y)
# get importance
importance = model.coef_[0]
# summarize feature importance
for i,v in enumerate(importance):
	print('Feature: %0d, Score: %.5f' % (i,v))
# plot feature importance
pyplot.bar([x for x in range(len(importance))], importance)
pyplot.show()
```

We can just make use of something like this, no?

Also, we will need to implement some plots for the feature importances. Would you mind creating an issue for this?

File: augurpy/calculate_AUC.py (hunk 6 above)
Comment (antschum) [reply augurpy/calculate_AUC.py:87]: In permut mode we want to shuffle the cell labels, so that each cell will be assigned a random label. I f we don't reset the index, then then the labels will just be resorted and matched to the original cell later on because of matching indexes.

**Comment by antschum** [comment]
> @antschum think you missed two comments of mine. (You can scroll through the changed files.) But besides this it looks great! Feel free to merge it after you've addressed my comments.

Ah! Im sorry, I totally missed those. I changed the variable name and replied to the other comment. 

> Regarding the feature importances:
> 
> ```
> # logistic regression for feature importance
> from sklearn.datasets import make_classification
> from sklearn.linear_model import LogisticRegression
> from matplotlib import pyplot
> # define dataset
> X, y = make_classification(n_samples=1000, n_features=10, n_informative=5, n_redundant=5, random_state=1)
> # define the model
> model = LogisticRegression()
> # fit the model
> model.fit(X, y)
> # get importance
> importance = model.coef_[0]
> # summarize feature importance
> for i,v in enumerate(importance):
> 	print('Feature: %0d, Score: %.5f' % (i,v))
> # plot feature importance
> pyplot.bar([x for x in range(len(importance))], importance)
> pyplot.show()
> ```
> 
> We can just make use of something like this, no?

So would you suggest just using the coefficient values as feature importances for the logistic regression? In the original they referenced the agresti method [here](https://think-lab.github.io/d/205/). I have not really found much on this, although in the blog post, it seems they are just subtracting the mean from each coefficient and then dividing by the standard deviation, right? 

> Also, we will need to implement some plots for the feature importances. Would you mind creating an issue for this?

Done! #42

**Comment by Zethson** [comment]
Ah, wasn't aware of this. Yes, I read the blog the same way.
Basically, this part:

> 
An alternative comparison of effects of quantitative predictors having different units uses standardized coefficients. The model is fitted to standardized predictors, replacing each xj by (xj−x¯j)/sxj. A 1-unit change in the standardized predictor is a standard deviation change in the original predictor. Then, each regression coefficient represents the effect of a standard deviation change in a predictor, controlling for the other variables. The standardized estimate for predictor xj is the unstandardized estimate β^j multiplied by sxj.

Feel free to merge it now :)
I'd suggest to implement the other things in follow up PRs.

**Comment by antschum** [comment]
> Ah, wasn't aware of this. Yes, I read the blog the same way. Basically, this part:
> 
> > 
> 
> An alternative comparison of effects of quantitative predictors having different units uses standardized coefficients. The model is fitted to standardized predictors, replacing each xj by (xj−x¯j)/sxj. A 1-unit change in the standardized predictor is a standard deviation change in the original predictor. Then, each regression coefficient represents the effect of a standard deviation change in a predictor, controlling for the other variables. The standardized estimate for predictor xj is the unstandardized estimate β^j multiplied by sxj.
> 
> Feel free to merge it now :) I'd suggest to implement the other things in follow up PRs.

Ah, I just read this and had already pushed - maybe you could take a quick look? I added some rough tests for the concordance correlation coefficient, so it should be more or less okay. And then we merge? My bad. 
Doing the other things in followup PRs :)

**Comment by antschum** [comment]
Forgot to run nox. 🤦‍♀️

File: tests/test_cv.py [hunk 16]
```diff
@@ -12,11 +12,24 @@
 # subsample 100 obs
 sc.pp.subsample(adata, n_obs=100, random_state=42)
 
-rf_estimator = create_estimator("random_forest_classifier", Params(random_state=42))
+rf_classifier = create_estimator("random_forest_classifier", Params(random_state=42))
+lr_classifier = create_estimator('logistic_regression_classifier')
```
Comment (Zethson) [inline tests/test_cv.py]: Doesn't this also have a random state? The sklearn implementation has one I thought?

**Review by Zethson — APPROVED** [review]
LGTM!

File: tests/test_cv.py (hunk 16 above)
Comment (antschum) [reply tests/test_cv.py]: True! adding it.
